### ISSUE TITLE
Committer queue is not fully processed

### ISSUE THREAD

**Issue by jsteggink** [Norconex/committer-core#16]
The committer queue is not fully processed because it's capped by the queueSize property. Since the queue can be bigger than the queueSize and is only called after a commit, the file queue grows and grows.

[https://github.com/Norconex/committer-core/blob/master/norconex-committer-core/src/main/java/com/norconex/committer/core/AbstractFileQueueCommitter.java#L175](url)

**Comment by essiembre** [comment]
When using ``AbstractFileQueueCommitter`` directly, the commit will only becalled at the end like you mention, unless you call it yourself more frequently.  If you want to be called after X number of documents, have a look at the ``AbstractBatchCommitter`` subclass, which does it for you.

Does that address your issue?

**Comment by essiembre** [comment]
If it can't keep up right now, you may have to slow it down, unfortunately.  I agree the queue could be improved and I am already sold to the idea of being able to supply your own queue.  I am marking this as a feature request.

The Committers will be seriously revisited in the next major version and something like RxJava will be given consideration.  Have you used RxJava in a few projects yourself?  Any examples?

**Comment by jsteggink** [comment]
I have some experience with [Reactor](https://projectreactor.io), which is another Reactive Streams framework. It's also used by the Spring framework. It would be my first choice as it's targeted to Java 8 and easily integrates with Kafka and RabbitMQ.

**Comment by truezjz** [comment]
Hi Pascal,

I'm also facing this issue,  after crawling 174K files, there are 12000 files left in the commiter-queue folder not processed.
Understand the issue will be addressed in next version,  is there a walk around for the time being?

**Comment by jsteggink** [comment]
Last year I fixed it. I can come up with a pull request tomorrow.

**Comment by truezjz** [comment]
thanks for for prompt response Jeroen, I will check it out,  that will be in the commiter-core right?

**Comment by jsteggink** [comment]
Yes, it's the committer-core. I need a little bit of extra time to make some unit tests and I have some stuff I haven't committed yet in my fork. In the meantime you can take a look of what I did: https://github.com/jsteggink/committer-core
I have added a reactive committer and a persistent queue based on RocksDb. This makes the committer-core more stable, way faster and potentially more scalable. In the future different persistent queue implementations could be added.

**Comment by truezjz** [comment]
Hi Jeroen, Thanks for the update, as you mentioned current code in your fork is not your final submit, looking forward to final submit;
Meanwhile as you mentioned, for temporary solution, I can set the queuesize in commiter to unlimited number, but the disadvantage is, it need large storage to hold the committer queue, correct?

**Comment by truezjz** [comment]
@jsteggink Jeroen any update?  if need I can help with the test.
